### ISSUE TITLE
Fast typing on set-url not registering last letter

### ISSUE THREAD

**Issue by noaccOS** [atlas-engineer/nyxt#1280]
**Describe the bug**
When writing fast, the last letter is often skipped

https://user-images.githubusercontent.com/24324352/113478142-3d9b0f00-9487-11eb-8f25-aa49638bf14e.mp4

**Precise recipe to reproduce the issue**
1. Type fast
2. Press enter to search
3. Last letter is probably skipped

**Information**
- OS name+version: Arch Linux
- Graphics card and driver: Nvidia gtx 1070 proprietary drivers
- Desktop environment / Window manager name+version: Xmonad git
- How you installed Nyxt (Guix pack, package manager, build from source): -git package from AUR
- Information from command copy-system-information:
Nyxt version: dont-use-this-tag2-1-g6d3a04e8 
Operating system kernel: Linux 5.11.11-zen1-1-zen
Lisp implementation: SBCL 2.1.1
Features: (WEBKIT2 WEBKIT2-2.32 WEBKIT2-TRACKING WEBKIT2-MUTE WEBKIT2-EMOJI
           WEBKIT2-MEDIA WEBKIT2-SANDBOXING GTK-3-22 GTK-3-20 GTK-3-18 GTK-3-16
           GTK-3-14 GTK-3-12 GTK-3-10 GTK-3-8 GTK-3-6 GTK-3-4 GTK GDK-3-22
           GDK-3-20 GDK-3-18 GDK-3-16 GDK-3-14 GDK-3-12 GDK-3-10 GDK-3-8
           GDK-3-6 GDK-3-4 CAIRO-1-10 CAIRO-1-12 GDK-PIXBUF GLIB-2-30 GLIB-2-32
           GLIB-2-34 GLIB-2-36 GLIB-2-38 GLIB-2-40 GLIB-2-42 GLIB-2-44
           GLIB-2-46 GLIB-2-48 GLIB-2-50 GLIB-2-52 GLIB-2-54 GLIB-2-56
           GLIB-2-58 GLIB FSET-EXT-STRINGS CUSTOM-HASH-TABLE-NATIVE SWANK
           PLUMP-UTF-32 GLOBAL-VARS DECLARE-TYPES PARENSCRIPT NAMED-READTABLES
           21BIT-CHARS CHUNGA FLEXI-STREAMS CLOSER-MOP CL-PPCRE-UNICODE
           CL-UNICODE CL-PPCRE BORDEAUX-THREADS SPLIT-SEQUENCE GRAY-STREAMS
           FLAT-NAMESPACE X86-64 UNIX CFFI FLAT-NAMESPACE SEQUENCE-EMPTYP
           FAST-IO-SV FAST-IO SBCL-USES-SB-ROTATE-BYTE ASDF-SYSTEM-CONNECTIONS
           CL-JSON-CLOS CL-JSON THREAD-SUPPORT QUICKLISP ASDF3.3 ASDF3.2
           ASDF3.1 ASDF3 ASDF2 ASDF OS-UNIX NON-BASE-CHARS-EXIST-P ASDF-UNICODE
           X86-64 GENCGC 64-BIT ANSI-CL COMMON-LISP ELF IEEE-FLOATING-POINT
           LINUX LITTLE-ENDIAN PACKAGE-LOCAL-NICKNAMES SB-CORE-COMPRESSION
           SB-LDB SB-PACKAGE-LOCKS SB-THREAD SB-UNICODE SBCL UNIX)

**Output when started from a shell**
<INFO> [14:17:52] Listening to socket "/home/noaccos/.local/share/nyxt/nyxt.socket".
Nyxt version dont-use-this-tag2-1-g6d3a04e8
<INFO> [14:17:52] Loading Lisp file "/home/noaccos/.config/nyxt/auto-config.lisp".
<INFO> [14:17:52] Loading Lisp file "/home/noaccos/.config/nyxt/init.lisp".
Gtk-Message: 14:17:52.422: Failed to load module "appmenu-gtk-module"
Gtk-Message: 14:17:52.620: Failed to load module "appmenu-gtk-module"
Gtk-Message: 14:17:52.631: Failed to load module "appmenu-gtk-module"
Gtk-Message: 14:17:52.710: Failed to load module "appmenu-gtk-module"
Gtk-Message: 14:17:52.753: Failed to load module "appmenu-gtk-module"
<INFO> [14:18:02] Loading history of 138 URLs from "/home/noaccos/.local/share/nyxt/history/default.lisp".
Gtk-Message: 14:18:05.613: Failed to load module "appmenu-gtk-module"
<INFO> [14:18:05] Loading "https://duckduckgo.com/?q=ab".
<INFO> [14:18:06] Finished loading "https://duckduckgo.com/?q=ab&ia=web".

**Comment by Ambrevar** [comment]
Good catch!

The error is with the prompt buffer not waiting for the end of the
calculation.

Should be easy to fix, I'll see to it next week.

**Comment by Ambrevar** [comment]
I suspect that a8dbb187529a012195abb0513cddf89fa0277f53 fixes it.
@noaccOS Can you confirm?

**Comment by noaccOS** [comment]
> I suspect that [a8dbb18](https://github.com/atlas-engineer/nyxt/commit/a8dbb187529a012195abb0513cddf89fa0277f53) fixes it.
> @noaccOS Can you confirm?

I'm sorry but I'm kind of having trouble building nyxt lately
It used to compile from AUR just fine but then it started building fine, but the executable was nowhere to be found (tried with `yay`, `paru` and manually with `makepkg`), and now it doesn't even compile anymore

I tried compiling from the git with `make all` but it gives the same errors as the AUR

I have this error just after `You already have the latest version of "quicklisp": 2021-04-11.`:

```
Unhandled ASDF/FIND-COMPONENT:MISSING-DEPENDENCY in thread #<SB-THREAD:THREAD "main thread" RUNNING
                                                              {10018882F3}>:
  Component "nyxt" not found, required by #<SYSTEM "nyxt/quicklisp">

Backtrace for: #<SB-THREAD:THREAD "main thread" RUNNING {10018882F3}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK Component "nyxt" not found, required by #<SYSTEM "nyxt/quicklisp"> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* Component "nyxt" not found, required by #<SYSTEM "nyxt/quicklisp">)
2: (INVOKE-DEBUGGER Component "nyxt" not found, required by #<SYSTEM "nyxt/quicklisp">)
3: (ERROR ASDF/FIND-COMPONENT:MISSING-DEPENDENCY :REQUIRED-BY #<ASDF/SYSTEM:SYSTEM "nyxt/quicklisp"> :REQUIRES "nyxt")
4: (ASDF/FIND-COMPONENT:RESOLVE-DEPENDENCY-NAME #<ASDF/SYSTEM:SYSTEM "nyxt/quicklisp"> "nyxt" NIL)
5: (ASDF/PLAN:MAP-DIRECT-DEPENDENCIES #<ASDF/LISP-ACTION:COMPILE-OP > #<ASDF/SYSTEM:SYSTEM "nyxt/quicklisp"> #<FUNCTION (LAMBDA (ASDF/PLAN::DEP-O ASDF/PLAN::DEP-C) :IN ASDF/PLAN:REDUCE-DIRECT-DEPENDENCIES) {100371372B}>)
6: (ASDF/PLAN:REDUCE-DIRECT-DEPENDENCIES #<ASDF/LISP-ACTION:COMPILE-OP > #<ASDF/SYSTEM:SYSTEM "nyxt/quicklisp"> #<FUNCTION (LAMBDA (DO ASDF/PLAN::DC ASDF/PLAN::STATUS) :IN ASDF/ACTION:COMPUTE-ACTION-STAMP) {10037136DB}> #<ASDF/PLAN:ACTION-STATUS :BITS 4 :STAMP 3828084169 :LEVEL 0 :INDEX NIL>)
7: ((:METHOD ASDF/ACTION:COMPUTE-ACTION-STAMP (T ASDF/OPERATION:OPERATION ASDF/COMPONENT:COMPONENT)) NIL #<ASDF/LISP-ACTION:COMPILE-OP > #<ASDF/SYSTEM:SYSTEM "nyxt/quicklisp"> :JUST-DONE T) [fast-method]
8: ((:METHOD ASDF/ACTION:MARK-OPERATION-DONE (ASDF/OPERATION:OPERATION ASDF/COMPONENT:COMPONENT)) #<ASDF/LISP-ACTION:COMPILE-OP > #<ASDF/SYSTEM:SYSTEM "nyxt/quicklisp">) [fast-method]
9: ((SB-PCL::EMF ASDF/ACTION:PERFORM) #<unused argument> #<unused argument> #<ASDF/LISP-ACTION:COMPILE-OP > #<ASDF/SYSTEM:SYSTEM "nyxt/quicklisp">)
10: ((LAMBDA NIL :IN ASDF/ACTION:CALL-WHILE-VISITING-ACTION))
11: ((:METHOD ASDF/ACTION:PERFORM-WITH-RESTARTS :AROUND (T T)) #<ASDF/LISP-ACTION:COMPILE-OP > #<ASDF/SYSTEM:SYSTEM "nyxt/quicklisp">) [fast-method]
12: ((:METHOD ASDF/PLAN:PERFORM-PLAN (T)) #<ASDF/PLAN:SEQUENTIAL-PLAN {1002D12213}>) [fast-method]
13: ((FLET SB-C::WITH-IT :IN SB-C::%WITH-COMPILATION-UNIT))
14: ((:METHOD ASDF/PLAN:PERFORM-PLAN :AROUND (T)) #<ASDF/PLAN:SEQUENTIAL-PLAN {1002D12213}>) [fast-method]
15: ((:METHOD ASDF/OPERATE:OPERATE (ASDF/OPERATION:OPERATION ASDF/COMPONENT:COMPONENT)) #<ASDF/LISP-ACTION:LOAD-OP > #<ASDF/SYSTEM:SYSTEM "nyxt/quicklisp"> :PLAN-CLASS NIL :PLAN-OPTIONS NIL) [fast-method]
16: ((SB-PCL::EMF ASDF/OPERATE:OPERATE) #<unused argument> #<unused argument> #<ASDF/LISP-ACTION:LOAD-OP > #<ASDF/SYSTEM:SYSTEM "nyxt/quicklisp">)
17: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
18: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) #<ASDF/LISP-ACTION:LOAD-OP > #<ASDF/SYSTEM:SYSTEM "nyxt/quicklisp">) [fast-method]
19: ((SB-PCL::EMF ASDF/OPERATE:OPERATE) #<unused argument> #<unused argument> ASDF/LISP-ACTION:LOAD-OP :NYXT/QUICKLISP)
20: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
21: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) ASDF/LISP-ACTION:LOAD-OP :NYXT/QUICKLISP) [fast-method]
22: (ASDF/SESSION:CALL-WITH-ASDF-SESSION #<FUNCTION (LAMBDA NIL :IN ASDF/OPERATE:OPERATE) {1002CF735B}> :OVERRIDE T :KEY NIL :OVERRIDE-CACHE T :OVERRIDE-FORCING NIL)
23: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
24: (ASDF/SESSION:CALL-WITH-ASDF-SESSION #<FUNCTION (LAMBDA NIL :IN ASDF/OPERATE:OPERATE) {1002BB901B}> :OVERRIDE NIL :KEY NIL :OVERRIDE-CACHE NIL :OVERRIDE-FORCING NIL)
25: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) ASDF/LISP-ACTION:LOAD-OP :NYXT/QUICKLISP) [fast-method]
26: (ASDF/OPERATE:LOAD-SYSTEM :NYXT/QUICKLISP)
27: (SB-INT:SIMPLE-EVAL-IN-LEXENV (ASDF/OPERATE:LOAD-SYSTEM :NYXT/QUICKLISP) #<NULL-LEXENV>)
28: (SB-INT:SIMPLE-EVAL-IN-LEXENV (WHEN (STRING= "true" "true") (ASDF/OPERATE:LOAD-SYSTEM :NYXT/QUICKLISP)) #<NULL-LEXENV>)
29: (EVAL (WHEN (STRING= "true" "true") (ASDF/OPERATE:LOAD-SYSTEM :NYXT/QUICKLISP)))
30: (SB-IMPL::PROCESS-EVAL/LOAD-OPTIONS ((:EVAL . "(require \"asdf\")") (:EVAL . "(asdf:load-asd \"/home/noaccos/.cache/yay/nyxt-browser-git/src/nyxt-browser//nyxt.asd\")") (:EVAL . "(when (string= \"true\" \"true\") (asdf:load-system :nyxt/quicklisp))") (:EVAL . "(ql:quickload :nyxt/gtk-application)") (:EVAL . "(asdf:make :nyxt/gtk-application)") (:EVAL . "(uiop:quit)") (:QUIT)))
31: (SB-IMPL::TOPLEVEL-INIT)
32: ((FLET SB-UNIX::BODY :IN SB-IMPL::START-LISP))
33: ((FLET "WITHOUT-INTERRUPTS-BODY-1" :IN SB-IMPL::START-LISP))
34: (SB-IMPL::START-LISP)

unhandled condition in --disable-debugger mode, quitting
; 
; compilation unit aborted
;   caught 1 fatal ERROR condition
```

I'm sorry but do you have any idea? @Ambrevar

**Comment by Ambrevar** [comment]
Very odd.

This issue was introduced with the move away from Makefile to .asd.
I'll look into it.

**Comment by noaccOS** [comment]
I tried again after some time and was able to build it successfully using `make all && sudo make install` (the AUR package is still giving the same error)

The problem does not seem solved on my end, it seems like there is still a delay between what you type and the 'URL or new query' text. The search always returns the results of the text there.

![image](https://user-images.githubusercontent.com/24324352/116704254-7ddca700-a9cb-11eb-97c2-d765aa36d1c1.png)

**Comment by Ambrevar** [comment]
Someone reported the build issue on the AUR: https://aur.archlinux.org/pkgbase/nyxt-browser-git/

I can't reproduce on Guix, maybe it's a current-working-directory issue?  I'll see if I can work it out in the Ubuntu pipeline.

**Comment by Ambrevar** [comment]
> The problem does not seem solved on my end, it seems like there is still a delay between what you type and the 'URL or new query' text. The search always returns the results of the text there.

I don't experience any delay nor can I reproduce.
If I type "abc" super fast, I instantly get "abc" in the "New URL or search query" source.

I see you are using Nvidia.  Did you disable compositing to get Nyxt to display properly?  Maybe that's the issue: the prompt buffer might be extra slow on your system due to this rendering issue.

In any case, if the source fails to update properly on slow hardware, it proves that there is a race condition.  I'll work it out.

**Comment by Ambrevar** [comment]
Ha, I could just reproduce it, but sadly only once!

I started Nyxt from the REPL, and exceptionnally all prompt buffers were lagging for about 1s before showing up.

If during this lag I would type "abc", then the prompt buffer would show "abc" in the input but not in the new URL source.  Just like on your screenshot, @noaccOS. 

At this point, if I force-refresh the buffer, the new URL source now correctly displays "abc".

So it may be that a source update in hanging somehow.

I tried reproducing this by introducing arbitrary `(sleep 1)` here and there in the library/prompter code base, to no avail.

**Comment by noaccOS** [comment]
Yes, I usually don't use a compositor but I tried enabling picom and the effect is still the same. I also wouldn't consider my setup "slow hardware" since I have a ryzen 3700X

I tried recording a video and counting the frames (recording is at 60fps)

Letter appears on "Open URL":
- a: frame 0
- b: frame 6
- c: frame 15

Letter appears in "URL or new query":
- a & b: at the same time at frame 15
- c: frame 18

It's definitely not 1 second but it's still a lot of time when writing fast

**Comment by Ambrevar** [comment]
@noaccOS About the build issue, I've posted some comments on the AUR: https://aur.archlinux.org/packages/nyxt-browser-git/#news

Can you see if purging the yay / paru / whatever cache helps?
Do you have any idea why a manual `make all && sudo make install` worked and not `makepkg`?
Could you report the full output when it fails?

**Comment by Ambrevar** [comment]
> It's definitely not 1 second but it's still a lot of time when writing fast

Hmm... I can't even produce a lag here on my end.  I could produce it only once, ever.  Need to work out a recipe.

Out of curiosity, do you have many bookmarks or a huge history?

**Comment by noaccOS** [comment]
> Do you have any idea why a manual `make all && sudo make install` worked and not `makepkg`?

I don't have any experience with PKGBUILD, but I tried looking at some other PKGBUILDs of -git packages and I noticed that in `pkgver()` and `build()` they all wrote `${srcdir}` and `${_pkgname}`, even if in `package()` they only wrote `$srcdir` and `$_pkgname`

So I tried wrapping the variables in brackets and it actually compiled with `makepkg`

**Comment by Ambrevar** [comment]
Thanks a lot for trying.

I don't think that's it, but just in case, can you prefix the line `cd` line of `build()` with 

```sh
echo "DEBUG: '$srcdir/$_pkgname'"
```

(Mind the single quotes.)
Then reproduce the error and the success and compare both outputs.

I don't think this would reveal anything though.
My hunch is that it works when you repeat the build.

Can you confirm that running `makepkg` multiple times works (or not)?

**Comment by noaccOS** [comment]
You were right, running `makepkg` multiple times works, regardless of the brackets for the variables

**Comment by noaccOS** [comment]
> Out of curiosity, do you have many bookmarks or a huge history?

No I don't. I also tried removing `~/.local/share/nyxt` and the lag was still there (but maybe less, I should do more tries on a slightly bigger history)

Here are some results on the "clean install"

Test 1:
| Letter | Frame in `Open URL` | Frame in `New URL or search query` |
|--------|---------------------|------------------------------------|
| a      | 0                   | 4                                  |
| b      | 3                   | 8                                  |
| c      | 6                   | 8                                  |

Test 2:
| Letter | Frame in `Open URL` | Frame in `New URL or search query` |
|--------|---------------------|------------------------------------|
| a      | 0                   | 4                                  |
| b      | 5                   | 10                                 |
| c      | 10                  | 13                                 |

Test 3:
| Letter | Frame in `Open URL` | Frame in `New URL or search query` |
|--------|---------------------|------------------------------------|
| a      | 0                   | 2                                  |
| b      | 1                   | 2                                  |
| c      | 3                   | 5                                  |

**Comment by Ambrevar** [comment]
Thanks for the thorough trials! :)

If you are recoding at 60 FPS, this means that `New URL or search query` has a delay of about 0.07-0.08 seconds.

I'll try to measure myself and see if I can reproduce this somehow.

**Comment by noaccOS** [comment]
I just noticed the recording was played at 25 fps because that seems to be the default for kdenlive

I recounted the frames and these are the actual results at 60fps

Test 1:
| Letter | Frame in `Open URL` | Frame in `New URL or search query` |
|--------|---------------------|------------------------------------|
| a      | 0                   | 9                                  |
| b      | 7                   | 18                                  |
| c      | 13                   | 18                                  |

Test 2:
| Letter | Frame in `Open URL` | Frame in `New URL or search query` |
|--------|---------------------|------------------------------------|
| a      | 0                   | 9                                  |
| b      | 13                   | 23                                 |
| c      | 24                  | 32                                 |

Test 3:
| Letter | Frame in `Open URL` | Frame in `New URL or search query` |
|--------|---------------------|------------------------------------|
| a      | 0                   | 6                                  |
| b      | 3                   | 6                                  |
| c      | 8                   | 12                                  |

They average at around a 7 frames delay which is about 0.12 seconds

**Comment by Ambrevar** [comment]
I can reproduce the slowdown!  If I go to duckduckgo with Javascript enabled, WebKitGTK gets very slow, it has about 1s of lag:

https://duckduckgo.com/?q=foo&ia=web

It does not happen on GitHub, so it must be a specific piece of Javascript code that's killing WebKitGTK.

We should probably report upstream.

Anyways, back to topic, now that I have a recipe to reproduce I'll be able to fix this race condition.

**Comment by eihli** [comment]
Just wanted to leave a note for visibility and so I can track this thread. I'm consistently experiencing this with Arch, Nvidia 2080 with proprietery drivers, StumpWM. 

```
Nyxt version: 2.0.0-106-g62bfacef
Renderer version: GI-GTK
Operating system kernel: Linux 5.10.41-1-MANJARO
Lisp implementation: SBCL 2.1.4
Features: (SLYNK WEBKIT2 WEBKIT2-2.32 WEBKIT2-PASTE-PLAINTEXT WEBKIT2-TRACKING
 WEBKIT2-MUTE WEBKIT2-EMOJI WEBKIT2-MEDIA WEBKIT2-SANDBOXING GTK-3-22 GTK-3-20
 GTK-3-18 GTK-3-16 GTK-3-14 GTK-3-12 GTK-3-10 GTK-3-8 GTK-3-6 GTK-3-4 GTK
 GDK-3-22 GDK-3-20 GDK-3-18 GDK-3-16 GDK-3-14 GDK-3-12 GDK-3-10 GDK-3-8 GDK-3-6
 GDK-3-4 CAIRO-1-10 CAIRO-1-12 GDK-PIXBUF GLIB-2-30 GLIB-2-32 GLIB-2-34
 GLIB-2-36 GLIB-2-38 GLIB-2-40 GLIB-2-42 GLIB-2-44 GLIB-2-46 GLIB-2-48
 GLIB-2-50 GLIB-2-52 GLIB-2-54 GLIB-2-56 GLIB-2-58 GLIB NYXT-2 FSET-EXT-STRINGS
 CUSTOM-HASH-TABLE-NATIVE SWANK PLUMP-UTF-32 GLOBAL-VARS DECLARE-TYPES
 PARENSCRIPT NAMED-READTABLES 21BIT-CHARS CHUNGA FLEXI-STREAMS CLOSER-MOP
 CL-PPCRE-UNICODE CL-UNICODE CL-PPCRE BORDEAUX-THREADS SPLIT-SEQUENCE
 GRAY-STREAMS FAST-IO-SV FAST-IO SBCL-USES-SB-ROTATE-BYTE
 ASDF-SYSTEM-CONNECTIONS CL-JSON-CLOS CL-JSON THREAD-SUPPORT FLAT-NAMESPACE
 X86-64 UNIX CFFI FLAT-NAMESPACE SEQUENCE-EMPTYP QUICKLISP ASDF3.3 ASDF3.2
 ASDF3.1 ASDF3 ASDF2 ASDF OS-UNIX NON-BASE-CHARS-EXIST-P ASDF-UNICODE X86-64
 GENCGC 64-BIT ANSI-CL COMMON-LISP ELF IEEE-FLOATING-POINT LINUX LITTLE-ENDIAN
 PACKAGE-LOCAL-NICKNAMES SB-LDB SB-PACKAGE-LOCKS SB-THREAD SB-UNICODE SBCL UNIX)

ASDF version: 3.3.1
Critical dependencies: (/home/eihli/src/nyxt/_build/submodules/cl-cffi-gtk/gtk/cl-cffi-gtk.asd
 /home/eihli/src/nyxt/_build/quicklisp-client/dists/quicklisp/software/cl-gobject-introspection-20210124-git/cl-gobject-introspection.asd
 /home/eihli/src/nyxt/_build/submodules/cl-webkit/webkit2/cl-webkit2.asd)

Quicklisp dist version: 2021-05-31
Quicklisp client version: 2020-01-04
Local project directories: (/home/eihli/src/nyxt/_build/submodules/
 /home/eihli/src/nyxt/_build/quicklisp-client/local-projects/)
Critical dependencies(#<SYSTEM cl-cffi-gtk / cl-cffi-gtk-20201220-git / quicklisp 2021-05-31>
 #<SYSTEM cl-gobject-introspection / cl-gobject-introspection-20210124-git / quicklisp 2021-05-31>
 #<SYSTEM cl-webkit2 / cl-webkit-20210531-git / quicklisp 2021-05-31>)
```

**Comment by noaccOS** [comment]
I think I know what the problem is

The 'URL or new query' field is updated only on key release, which make it feel slow

**Comment by Ambrevar** [comment]
> The 'URL or new query' field is updated only on key release, which make it feel slow

This was not the original post issue, but still an important issue.
I've fixed it with 7ff39a04dfa438ea27f9d149c31a793770bf5144.

To test it, keep pressing say "a" in the prompt buffer input: it should
now display a sequence of "a" progressively, instead of displaying it on
release only.

I believe I've fixed the original post issue in
7ff39a04dfa438ea27f9d149c31a793770bf5144.
The trick is to refresh the prompt buffer when we see that the HTML and
prompter inputs are out-of-sync (apparently this happens due to a
synchronization issue with WebKitGTK).

Can anyone confirm this solves all the issues in this thread?
